File: agents/python-agent/hmdev/messaging/agent/security/my_security.py
import base64
import hmac, hashlib
import json
import logging

from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives import hashes

from hmdev.messaging.agent.security.aes_ctr import AesCtr

logger = logging.getLogger(__name__)


class MySecurity:
    """Static-like utility class for encryption, signing, and verification."""

    def __init__(self):
        raise RuntimeError("No instances allowed")

    @staticmethod
    def encrypt_and_sign(message: str, key: str) -> str:
        data = {
            "cipher": MySecurity.encrypt(message, key),
            "hash": MySecurity.hash(message, key),
        }
        return json.dumps(data)

    @staticmethod
    def decrypt_and_verify(cipher_msg_str: str, key: str) -> str | None:
        try:
            cipher_msg = json.loads(cipher_msg_str)
            message = MySecurity.decrypt(cipher_msg.get("cipher", ""), key)

            if message is None:
                return None

            if MySecurity.hash(message, key).strip() != cipher_msg.get("hash", "").strip():
                return None
            return message
        except Exception as e:
            logger.debug(f"decrypt_with_md5_auth error: {e}")
            return None

    # ------------------ AES CTR helper ------------------

    @staticmethod
    def encrypt(plain: str, key: str) -> str:
        try:
            return AesCtr.encrypt(plain, key, 128)
        except Exception as e:
            logger.debug(f"encrypt error: {e}")
            return ""

    @staticmethod
    def decrypt(cipher: str, key: str) -> str | None:
        try:
            return AesCtr.decrypt(cipher, key, 128)
        except Exception as e:
            logger.debug(f"decrypt error: {e}")
            return None

    # ------------------ Key derivation ------------------

    @staticmethod
    def derive_channel_secret(channel_name: str, password: str) -> str:
        combined = (channel_name + password).encode()
        salt = b"messaging-api"  # must match JS
        iterations = 100_000
        key_length = 32  # 256 bits

        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=key_length,
            salt=salt,
            iterations=iterations,
            backend=default_backend(),
        )
        key_bytes = kdf.derive(combined)
        return base64.b64encode(key_bytes).decode()

    # ------------------ MD5 Hash ------------------

    @staticmethod
    def hash(msg: str, key_str: str) -> str:
        try:
            # return hashlib.sha256(msg.encode("utf-8")).hexdigest()
            return hmac.new(key_str.encode("utf-8"), msg.encode("utf-8"), hashlib.sha256).hexdigest()
        except Exception as e:
            logger.warning(f"hash error: {e}")
            return None

File: agents/java-agent/src/main/java/com/hmdev/messaging/agent/security/MySecurity.java
package com.hmdev.messaging.agent.security;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.spec.KeySpec;
import java.util.Base64;
import java.util.Objects;

import org.apache.commons.io.IOUtils;
import org.json.JSONObject;

import com.hmdev.messaging.agent.security.aes.AesCtr;

import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.spec.SecretKeySpec;

public final class MySecurity {
    private MySecurity() { /* no instances */ }

    private static final Logger logger = LoggerFactory.getLogger(MySecurity.class);

    public static String encryptAndSign(String message, String key) {
        JSONObject $myObj = new JSONObject("{}");
        $myObj.put("cipher", encrypt(message, key));
        $myObj.put("hash", hash(message, key));
        return $myObj.toString();
    }

    public static String decryptAndVerify(String cipherMsgStr, String key) {

        try {
            JSONObject $cipherMsg = new JSONObject(cipherMsgStr);

            String message = decrypt($cipherMsg.optString("cipher"), key);

            if (!Objects.equals(hash(message, key), $cipherMsg.optString("hash"))) {
                return null;
            } else {
                return message;
            }
        } catch (Exception e) {
            logger.debug("decryptWithMd5Auth error: " + e.getMessage());
            return null;
        }

    }

    public static String encrypt(String $plain, String $key) {
        try {
            return AesCtr.encrypt($plain, $key, 128);
        } catch (Exception e) {
            logger.debug("encrypt error: " + e.getMessage());
            return "";
        }
    }

    public static String decrypt(String $cipher, String $key) {
        try {
            return AesCtr.decrypt($cipher, $key, 128);
        } catch (Exception e) {
            logger.debug("decrypt error: " + e.getMessage());
        }

        return null;

    }

    public static String deriveChannelSecret(String channelName, String password) throws Exception {
        String combined = channelName + password;
        byte[] salt = "messaging-api".getBytes(); // must match JS salt
        int iterations = 100_000;
        int keyLength = 256; // bits

        KeySpec spec = new PBEKeySpec(combined.toCharArray(), salt, iterations, keyLength);
        SecretKeyFactory factory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256");

        byte[] keyBytes = factory.generateSecret(spec).getEncoded();
        return Base64.getEncoder().encodeToString(keyBytes); // same format as JS
    }

    public static String blocksEncrypt(Cipher encryptor, String plain) {

        if (encryptor == null) {
            return plain;
        }
        try {
            StringBuilder cipherPayload = new StringBuilder();

            for (int i = 0; i < plain.length(); i += 200) {

                int upperLimit = i + 200;
                upperLimit = Math.min(upperLimit, plain.length());

                byte[] salted = encryptor.doFinal(plain.substring(i, upperLimit).getBytes());

                cipherPayload.append(new String(Base64.getEncoder().encode(salted)));
            }

            return cipherPayload.toString();
        } catch (Exception e) {
            logger.warn(e.getMessage());
            return "";
        }

    }

    public static String hash(String msg, String key) {
        if (msg == null || key == null) {
            return null;
        }
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            SecretKeySpec secretKeySpec = new SecretKeySpec(key.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
            mac.init(secretKeySpec);
            byte[] hmacBytes = mac.doFinal(msg.getBytes(StandardCharsets.UTF_8));
            return toHexString(hmacBytes); // keep your existing hex converter
        } catch (Exception e) {
            logger.warn(e.getMessage());
            return null;
        }
    }

    public static String toHexString(byte[] bytes) {
        StringBuilder hexString = new StringBuilder();

        for (byte aByte : bytes) {
            String hex = Integer.toHexString(0xFF & aByte);
            if (hex.length() == 1) {
                hexString.append('0');
            }
            hexString.append(hex);
        }

        return hexString.toString();

    }
}